src/mlpack/methods/decision_tree/best_binary_numeric_split_impl.hpp

```diff
@@ -185,14 +185,297 @@ double BestBinaryNumericSplit<FitnessFunction>::SplitIfBetter(
+    const ResponsesType& responses,
+    const WeightVecType& weights,
+    const size_t minimumLeafSize,
+    const double minimumGainSplit,
+    double& splitInfo,
+    AuxiliarySplitInfo& /* aux */)
+{
+  typedef typename ResponsesType::elem_type RType;
+  typedef typename WeightVecType::elem_type WType;
+
+  FitnessFunction fitnessFunction;
+
+  // First sanity check: if we don't have enough points, we can't split.
+  if (data.n_elem < (minimumLeafSize * 2))
+    return DBL_MAX;
+  if (bestGain == 0.0)
+    return DBL_MAX; // It can't be outperformed.
+
+  // Next, sort the data.
+  arma::uvec sortedIndices = arma::sort_index(data);
+  arma::Row<RType> sortedResponses(responses.n_elem);
+  arma::Row<WType> sortedWeights;
+  for (size_t i = 0; i < sortedResponses.n_elem; ++i)
+    sortedResponses[i] = responses[sortedIndices[i]];
+
+  // Sanity check: if the first element is the same as the last, we can't split
+  // in this dimension.
+  if (data[sortedIndices[0]] == data[sortedIndices[sortedIndices.n_elem - 1]])
+    return DBL_MAX;
+
+  // Only initialize if we are using weights.
+  if (UseWeights)
+  {
+    sortedWeights.set_size(sortedResponses.n_elem);
+    // The weights must keep the same order as the responses.
+    for (size_t i = 0; i < sortedResponses.n_elem; ++i)
+      sortedWeights[i] = weights[sortedIndices[i]];
+  }
+
+  double bestFoundGain = std::min(bestGain + minimumGainSplit, 0.0);
+  bool improved = false;
+  // Force a minimum leaf size of 1 (empty children don't make sense).
+  const size_t minimum = std::max(minimumLeafSize, (size_t) 1);
+
+  WType totalWeight = 0.0;
+  WType leftChildWeight = 0.0;
+  WType rightChildWeight = 0.0;
+
+  if (UseWeights)
+  {
+    totalWeight = arma::accu(sortedWeights);
+    bestFoundGain *= totalWeight;
+
+    for (size_t i = 0; i < minimum - 1; ++i)
+      leftChildWeight += sortedWeights[i];
+
+    for (size_t i = minimum - 1; i < data.n_elem; ++i)
+      rightChildWeight += sortedWeights[i];
+  }
+  else
+  {
+    bestFoundGain *= data.n_elem;
+  }
+
+  // Initialize and precompute various statistics to efficiently compute gain
+  // values for all possible splits.
+  fitnessFunction.BinaryScanInitialize<UseWeights>(sortedResponses,
+      sortedWeights, minimum);
+
+  // Loop through all possible split points, choosing the best one.
+  for (size_t index = minimum; index < data.n_elem - minimum + 1; ++index)
+  {
+    if (UseWeights)
+    {
+      leftChildWeight += sortedWeights[index - 1];
+      rightChildWeight -= sortedWeights[index - 1];
+    }
+
+    // Steps through the current index and updates the cached data.
+    fitnessFunction.BinaryStep<UseWeights>(sortedResponses,
+        sortedWeights, index - 1);
+
+    // Make sure that the value has changed.
+    if (data[sortedIndices[index]] == data[sortedIndices[index - 1]])
+      continue;
+
+    // Calculate the gain for the left and right child.
+    auto binaryGains = fitnessFunction.BinaryGains();
```

Comment:
Would it be worthwhile to check via SFINAE for the existence of `BinaryGains()` also?

As a side comment, personally I think `auto` can obscure the meaning of the code a bit.  Even though it is a bit more verbose, I would consider going with the exact type definition for clarity:

```suggestion
    std::tuple<double, double> binaryGains = fitnessFunction.BinaryGains();
```